flytekit/core/type_engine.py

```diff
@@ -1679,22 +1677,53 @@ def get_dict_types(t: Optional[Type[dict]]) -> typing.Tuple[Optional[type], Opti
                         raise ValueError(
                             f"Flytekit does not currently have support for FlyteAnnotations applied to dicts. {t} cannot be parsed."
                         )
-            if _origin is dict and _args is not None:
+            if _origin in [dict, Annotated] and _args is not None:
                 return _args  # type: ignore
         return None, None
 
     @staticmethod
-    def dict_to_generic_literal(v: dict) -> Literal:
+    def dict_to_generic_literal(v: dict, allow_pickle: bool) -> Literal:
         """
         Creates a flyte-specific ``Literal`` value from a native python dictionary.
         """
-        return Literal(scalar=Scalar(generic=_json_format.Parse(json.dumps(v), _struct.Struct())))
+        from flytekit.types.pickle import FlytePickle
+
+        try:
+            return Literal(
+                scalar=Scalar(generic=_json_format.Parse(json.dumps(v), _struct.Struct())),
+                metadata={"format": "json"},
+            )
+        except TypeError as e:
+            if allow_pickle:
+                remote_path = FlytePickle.to_pickle(v)
+                return Literal(
+                    scalar=Scalar(
+                        generic=_json_format.Parse(json.dumps({"pickle_file": remote_path}), _struct.Struct())
+                    ),
+                    metadata={"format": "pickle"},
+                )
+            raise e
+
+    def is_pickle(self, python_type: Type[dict]) -> typing.Tuple[bool, Type]:
+        base_type, *metadata = self.extract_types_or_metadata(python_type)
+
+        for each_metadata in metadata:
+            if isinstance(each_metadata, OrderedDict):
+                allow_pickle = each_metadata.get("allow_pickle", False)
+                break
+
+        return allow_pickle, base_type
 
     def get_literal_type(self, t: Type[dict]) -> LiteralType:
         """
         Transforms a native python dictionary to a flyte-specific ``LiteralType``
         """
-        tp = self.get_dict_types(t)
+        if get_origin(t) is Annotated:
+            base_type, *_ = self.extract_types_or_metadata(t)
+            tp = get_args(base_type)
+        else:
+            tp = self.extract_types_or_metadata(t)
```

Comment:
nit: could we move the logic here into `self.extract_types_or_metadata`? something like

```python
def extract_types_or_metadata(self, python_type):
    if get_origin(python_type) is Annotated:
        python_type = get_args(python_type)[0]
    ... # original code here
```

so we can change the code here to
```python
tp = self.extract_types_or_metadata(t)
```

Reply:
i'm not sure if this is possible 'cause i don't want to always return the dict key and value types. i also need access to the annotation metadata to check if `allow_pickle` is set to `True`.

```python
def is_pickle(self, python_type: Type[dict]) -> typing.Tuple[bool, Type]:
        base_type, *metadata = self.extract_types_or_metadata(python_type)

        for each_metadata in metadata:
            if isinstance(each_metadata, OrderedDict):
                allow_pickle = each_metadata.get("allow_pickle", False)
                return allow_pickle, base_type

        return False, base_type
```

Reply:
so if `extract_types_or_metadata` returns the args, i can either fetch the metadata or the actual dict types depending on what i need. makes sense?

Reply:
i could probably encapsulate this logic in a separate function to avoid repetition. ok?